doc/reference/networking/mqtt.rst

```diff
@@ -1,11 +1,167 @@
 .. _mqtt_socket_interface:
 
 MQTT
-#####
+####
 
 Overview
 ********
 
+MQTT (Message Queuing Telemetry Transport) is an application layer protocol
+which works on top of the TCP/IP. It is desiged as an lightweight
+publish/subscribe messaging transport for machine-to-machine communication.
+For more information about the protocol itself, see http://mqtt.org/.
+
+Zephyr provides an MQTT client library built on top of BSD sockets API. The
+library is configurable at a per-client basis, with support for MQTT in versions
+3.1.0 and 3.1.1. The Zephyr MQTT implementation can be used with either plain
+sockets and communicate over TCP, or with secure sockets and communicate over
+TLS. See :ref:`bsd_sockets_interface` for more information about Zephyr sockets.
+
+MQTT clients require MQTT server to connect to. Such server, called an MQTT Broker,
+is responsible for managing client subcribtions and distributing messages
+published by clients. There are many implementations of MQTT brokers, one of them
+being Eclipse Mosquitto. See https://mosquitto.org/ for more information about
+the Eclipse Mosquitto project.
+
+Sample usage
+************
+
+To create an MQTT client, a client context structure and buffers need to be
+defined:
+
+.. code-block:: c
+
+   /* Buffers for MQTT client. */
+   static u8_t rx_buffer[256];
+   static u8_t tx_buffer[256];
+
+   /* MQTT client context */
+   static struct mqtt_client client_ctx;
+
+Multiple MQTT client instances can be created in the application and managed
+independently. Additionally, a structure for MQTT Broker address information
+needs to be defined. This structure have to be accessible throughout the lifespan
+of the MQTT client and can be shared among MQTT clients:
+
+.. code-block:: c
+
+   /* MQTT Broker address information. */
+   static struct sockaddr_storage broker;
+
+An MQTT client library will notify MQTT events to the application through a
+callback function. Therefore, one needs to be created to handle respective
+events:
+
+.. code-block:: c
+
+   void mqtt_evt_handler(struct mqtt_client *client,
+                         const struct mqtt_evt *evt)
+   {
+      switch (evt->type) {
+         /* Handle events here. */
+      }
+   }
+
+For a list of possible events, see :ref:`mqtt_api_reference`.
+
+In order to be usable, the client context structure needs to be initalized and
+set up. An example configuration for TCP transport is show below:
+
+.. code-block:: c
+
+   mqtt_client_init(&client_ctx);
+
+   /* MQTT client configuration */
+   client_ctx.broker = &broker;
+   client_ctx.evt_cb = mqtt_evt_handler;
+   client_ctx.client_id.utf8 = (u8_t *)"zephyr_mqtt_client";
+   client_ctx.client_id.size = strlen("zephyr_mqtt_client");
```

Comment:
It probably does not matter much for example code but it is better to write this

```
client_ctx.client_id.size = sizeof("zephyr_mqtt_client") - 1;
```
That way we avoid a function call as the sizeof() is evaluated at compile time.